core/internal/cltest/factories.go

```diff
@@ -633,13 +604,74 @@ func MustInsertFatalErrorEthTx(t *testing.T, store *strpkg.Store) models.EthTx {
 	return etx
 }
 
-func MustInsertRandomKey(t *testing.T, store *strpkg.Store) models.Key {
-	k := models.Key{Address: models.EIP55Address(NewAddress().Hex()), JSON: JSONFromString(t, `{"key": "factory"}`)}
-	require.NoError(t, store.CreateKeyIfNotExists(k))
-	return k
+func MustAddRandomKeyToKeystore(t testing.TB, store *strpkg.Store, opts ...interface{}) (models.Key, common.Address) {
+	t.Helper()
+
+	k := MustGenerateRandomKey(t, opts...)
+	MustAddKeyToKeystore(t, &k, store)
+	return k, k.Address.Address()
+}
+
+func MustAddKeyToKeystore(t testing.TB, key *models.Key, store *strpkg.Store) {
+	_, err := store.KeyStore.Import(key.JSON.Bytes(), Password, Password)
+	require.NoError(t, err)
+	require.NoError(t, store.DB.Create(key).Error)
+}
+
+// MustInsertRandomKey inserts a randomly generated (not cryptographically
+// secure) key for testing
+// If using this with the keystore, it should be called before the keystore loads keys from the database
+func MustInsertRandomKey(t testing.TB, db *gorm.DB, opts ...interface{}) models.Key {
+	t.Helper()
+
+	key := MustGenerateRandomKey(t, opts...)
+
+	require.NoError(t, db.Create(&key).Error)
+	return key
+}
+
+func MustGenerateRandomKey(t testing.TB, opts ...interface{}) models.Key {
+	privateKeyECDSA, err := ecdsa.GenerateKey(crypto.S256(), rand.Reader)
+	require.NoError(t, err)
+	id := pbormanuuid.NewRandom()
+	k := &keystore.Key{
+		Id:         id,
+		Address:    crypto.PubkeyToAddress(privateKeyECDSA.PublicKey),
+		PrivateKey: privateKeyECDSA,
+	}
+	keyjsonbytes, err := keystore.EncryptKey(k, Password, utils.FastScryptParams.N, utils.FastScryptParams.P)
+	require.NoError(t, err)
+	keyjson, err := models.ParseJSON(keyjsonbytes)
+	require.NoError(t, err)
+	eip, err := models.EIP55AddressFromAddress(k.Address)
+	require.NoError(t, err)
+
+	var nextNonce *int64
+	var funding bool
+	for _, opt := range opts {
+		switch v := opt.(type) {
+		case int:
+			i := int64(v)
+			nextNonce = &i
+		case int64:
+			nextNonce = &v
+		case bool:
+			funding = v
+		default:
+			t.Fatalf("unrecognised option type: %T", v)
```

Comment:
I know this is just for testing - but we it seems simpler to just pass nonce and funding as arguments to avoid this (check at compile time instead)? If you need to add a new option, a search and replace on the signature with a default for the new param isn't so bad

Reply:
Since nonce can be nil/pointer it makes it a bit annoying to use otherwise